File: Day 061/Problem 1472. Design Browser History.cpp
class BrowserHistory {
private:
    struct ListNode {
        string url;
        ListNode* next;
        ListNode* prev;

        ListNode(): url(""), next(nullptr), prev(nullptr) {}
        ListNode(string str): url(str), next(nullptr), prev(nullptr) {}
        ListNode(string str, ListNode* prevPtr, ListNode* nextPtr): url(str), next(nextPtr), prev(prevPtr) {}
    };
public:
    ListNode* history = nullptr;

    BrowserHistory(string homepage) {
        history = new ListNode(homepage);
    }
    
    void visit(string url) {
        ListNode* newPage = new ListNode(url);
        history->next = newPage;
        newPage->prev = history;

        history = newPage;
    }
    
    string back(int steps) {
        while (history && history->prev && steps) {
            history = history->prev;
            steps--;
        }

        return history->url;
    }
    
    string forward(int steps) {
        while (history && history->next && steps) {
            history = history->next;
            steps--;
        }

        return history->url;
    }
};

/**
 * Your BrowserHistory object will be instantiated and called as such:
 * BrowserHistory* obj = new BrowserHistory(homepage);
 * obj->visit(url);
 * string param_2 = obj->back(steps);
 * string param_3 = obj->forward(steps);
 */

File: Day 089/Problem 1472. Design Browser History.java
class DoublyListNode {
    String val;
    DoublyListNode prev;
    DoublyListNode next;
    DoublyListNode() {}
    DoublyListNode(String val) {this.val = val;}
    DoublyListNode(String val, DoublyListNode prev, DoublyListNode next) {this.val = val; this.prev = prev; this.next = next;}
}

class BrowserHistory {
    DoublyListNode node;
    public BrowserHistory(String homepage) {
        node = new DoublyListNode(homepage);
    }
    
    public void visit(String url) {
        DoublyListNode newNode = new DoublyListNode(url);
        newNode.prev = node;
        node.next = newNode;

        node = node.next;
    }
    
    public String back(int steps) {
        while (steps > 0 && node.prev != null) {
            node = node.prev;
            steps--;
        }

        return node.val;
    }
    
    public String forward(int steps) {
        while (steps > 0 && node.next != null) {
            node = node.next;
            steps--;
        }

        return node.val;
    }
}